File: 프로그래머스/unrated/150368. 이모티콘 할인행사/이모티콘 할인행사.cpp
#include<iostream>
#include <string>
#include <vector>

using namespace std;

int en, un;
int maxPlus, maxCost;
int order[7];
int user[100][2];
vector<int> emoticon;

void solve(){
    int plus = 0, cost = 0;
    for(int i = 0; i < un; i++){
        int sell = user[i][0];
        
        int sum = 0;
        for(int j = 0; j < en; j++){
            if(sell <= order[j]){
                sum += emoticon[j] * (100 - order[j]) / 100;
            }
        }
        
        if(sum >= user[i][1]){
            plus++;
        }else{
            cost += sum;
        }
    }
    
    if(plus > maxPlus){
        maxPlus = plus;
        maxCost = cost;
    }else if(plus == maxPlus && cost > maxCost){
        maxCost = cost;
    }
    
}

void combi(int cnt){
    if(cnt == en){
        solve();
        return;
    }
    
    for(int i = 1; i <= 4; i++){
        order[cnt] = i * 10;
        combi(cnt+1);
    }
}

vector<int> solution(vector<vector<int>> users, vector<int> emoticons) {
    vector<int> answer;
    
    en = emoticons.size();
    un = users.size();
    maxPlus = 0;
    maxCost = 0;
    
    for(int i = 0; i < un; i++){
        user[i][0] = users[i][0];
        user[i][1] = users[i][1];
    }
    
    for(int i = 0; i < en; i++){
        emoticon.push_back(emoticons[i]);
    }
    
    combi(0);
    
    answer.push_back(maxPlus);
    answer.push_back(maxCost);
    
    return answer;
}

File: 프로그래머스/2/150368. 이모티콘 할인행사/이모티콘 할인행사.java
import java.util.*;

class Solution {
    
    static int[] sale;
    static int[][] _users;
    static int[] _emoticons;
    static int[] answer;
    
        
    static void simul() {
        int count = 0;
        int totalSum = 0;
        
        for(int[] user : _users) {
            int sum = 0;
            for(int i = 0; i < _emoticons.length; ++i) {
                int tmp = _emoticons[i] * (100 - sale[i]) / 100;
                
                // 구매
                if(user[0] <= sale[i]) {
                    sum += tmp;
                }
            }
            
            if(sum >= user[1]) {
                count++;
            }
            else {
                totalSum += sum;
            }
        }
        
        if(count > answer[0] || (count == answer[0] && totalSum > answer[1])) {
            answer[0] = count;
            answer[1] = totalSum;
        }
    }
    
    static void combi(int idx) {
        if(idx == sale.length) {
            simul();
            return;
        }
        
        int k = 10;
        while(k <= 40) {
            sale[idx] = k;
            combi(idx + 1);
            k += 10;
        }
    }
    
    public int[] solution(int[][] users, int[] emoticons) {
        _users = users;
        _emoticons = emoticons;
        answer = new int[2];
        
        sale = new int[emoticons.length];
        combi(0);

        return answer;
    }
}